src/analysis/complex/abs_max.lean

```diff
@@ -0,0 +1,153 @@
+  { rintros ζ (hζ : abs (ζ - z) = r),
+    simpa [norm_smul, hζ, ← div_eq_inv_mul] using (div_le_div_right hr).2 (hz (hsub' hζ)) },
+  have hlt : ∥(w - z)⁻¹ • f w∥ < ∥f z∥ / r,
+    by simpa [norm_smul, ← div_eq_inv_mul] using (div_lt_div_right hr).2 hw_lt,
+  have : ∥∮ ζ in C(z, r), (ζ - z)⁻¹ • f ζ∥ < 2 * π * r * (∥f z∥ / r),
+    from circle_integral.norm_integral_lt_of_norm_le_const_of_lt hr hcont hle ⟨w, rfl, hlt⟩,
+  refine this.ne _,
+  rw circle_integral_sub_inv_smul_of_differentiable_on (mem_ball_self hr) (hd.mono hsub),
+  field_simp [norm_smul, hr.ne', abs_of_pos real.pi_pos],
+  ac_refl
+end
+
+/-- If `f : E → F` is complex differentiable on a set `s`, the norm of `f` takes it maximum on `s`
+at `z` and `w` is a point such that the closed ball with center `z` and radius `dist w z` is
+included in `s`, then `∥f w∥ = ∥f z∥`. -/
+lemma norm_eq_norm_of_is_max_on_of_closed_ball_subset {f : E → F} {s : set E} {z w : E}
+  (hd : differentiable_on ℂ f s) (hz : is_max_on (norm ∘ f) s z)
+  (hsub : closed_ball z (dist w z) ⊆ s) :
+  ∥f w∥ = ∥f z∥ :=
+begin
+  set e : ℂ → E := λ t : ℂ, t • (w - z) + z,
+  suffices : ∥(f ∘ e) 1∥ = ∥(f ∘ e) 0∥, by simpa [e],
+  have : differentiable_on ℂ (f ∘ e) ((λ t : ℂ, t • (w - z) + z) ⁻¹' s),
+    from hd.comp ((differentiable_on_id.smul_const (w - z)).add_const z) subset.rfl,
+  refine norm_eq_norm_of_is_max_on_of_closed_ball_subset_dim1
+    this _ _,
+  { refine λ t ht, _, simpa [e] using hz ht },
+  { refine λ t ht, hsub _,
+    have : abs t ≤ 1, by simpa using ht,
+    simpa [dist_eq_norm, norm_smul] using mul_le_of_le_one_left (norm_nonneg (w - z)) this }
+end
+
+lemma norm_eventually_eq_of_is_local_max {f : E → F} {c : E}
+  (hd : ∀ᶠ z in 𝓝 c, differentiable_at ℂ f z) (hc : is_local_max (norm ∘ f) c) :
+  ∀ᶠ y in 𝓝 c, ∥f y∥ = ∥f c∥ :=
+begin
+  rcases nhds_basis_closed_ball.eventually_iff.1 (hd.and hc) with ⟨r, hr₀, hr⟩,
+  exact nhds_basis_closed_ball.eventually_iff.2 ⟨r, hr₀, λ w hw,
+    norm_eq_norm_of_is_max_on_of_closed_ball_subset
+      (λ z hz, (hr hz).1.differentiable_within_at) (λ z hz, (hr hz).2)
+      (closed_ball_subset_closed_ball hw)⟩
+end
+
+lemma is_open_set_of_mem_nhds_and_is_max_on_norm {f : E → F} {s : set E}
+  (hd : differentiable_on ℂ f s) :
+  is_open {z | s ∈ 𝓝 z ∧ is_max_on (norm ∘ f) s z} :=
+begin
+  refine is_open_iff_mem_nhds.2 (λ z hz, (eventually_eventually_nhds.2 hz.1).and _),
+  replace hd : ∀ᶠ w in 𝓝 z, differentiable_at ℂ f w, from hd.eventually_differentiable_at hz.1,
+  exact (norm_eventually_eq_of_is_local_max hd $ (hz.2.is_local_max hz.1)).mono
+    (λ x hx y hy, le_trans (hz.2 hy) hx.ge)
+end
+
+/-- **Maximum modulus principle**: if `f : E → F` is complex differentiable on a nonempty compact
+set `K`, then there exists a point `z ∈ frontier K` such that `λ z, ∥f z∥` takes it maximum value on
+`K` at `z`. -/
+lemma exists_mem_frontier_is_max_on_norm [nontrivial E] {f : E → F} {K : set E} (hK : is_compact K)
```

Comment:
In this theorem and the following ones, it would be more satisfactory to assume only differentiability on the interior of K and continuity on K, as in many applications the assumptions are of this form. But if you want to postpone that to a later PR (or until it's needed), that's fine for me.